README.md
# NJ-TimePicker
<p align="center">
  <img src="https://i.ibb.co/cb6RTVS/12.png">
  <img src="https://i.ibb.co/ydRmPyb/242.png">
</p> 

NJ-TimePicker is a pure JavaScript based timepicker that enables you to select time both in 12-hour and 24-hour format in a web and mobile friendly user interface. Checkout the [DEMO HERE](https://nj-coder.com/nj-libs/nj-timepicker/index.html) or [FIDDLE IT OUT](https://jsfiddle.net/njcoder/5dc0trb9/)

## Table of Contents

- [Installation](#installation)
- [Configuration](#configuration)
- [Usage](#usage)
- [How to Contribute](#how-to-contribute)
- [License](#license)

## Installation
```
npm install --save nj-timepicker
```

You can use NJ-TimePicker as an ES6 module as follows:

```javascript
import NJTimePicker from 'nj-timepicker'; 

const picker = new NJTimePicker(...);
```

Alertnatively you can include the `njtimepicker.min.js` script before the closing `</body>` tag and then in your JS create a new instance of NJ-TimePicker as below.

```html
<script src="path/to/njtimepicker.min.js"></script>
<script>
    const picker = new NJTimePicker(...);
</script>
```
## Usage
```js
let container = document.querySelector('#container');
// init picker
var picker = new NJTimePicker({
    targetID: 'container'
});
// on save event
picker.on('save', function (data) {
    container.textContent = data.fullResult;
});
```

## Configuration
The plugin ships with various configuration and events that can be used to meet your requirements.

### Initialization Options
| Property           | Values             | Default                  | Description  |
| :----------------- |:-------------------| :------------------------| :------------|
| target             | DOM Element        | --                       | target DOM element |
| targetID           | String             | --                       | the DOM element ID |
| id                 | String             | ```alphanumeric string```| a unique ID for the picker |
| format             | ```12 \| 24```     | ```12```                 | picker display format |
| clickOutsideToClose| ```true \| false```| ```true```               | to close the picker when clicked outside |
| autoSave           | ```true \| false```| ```false```               | save when all values are selected |
| minutes            | `array`            | `[0, 15, 30, 45]`        | a list of minutes |
| disabledHours      | `array`            | `[]`                     | a list of disabled Hours |
| disabledMinutes    | `array`            | `[]`                     | a list of disabled Minutes |
| texts              | `object`           | `{`<br>`header: ''`<br>`hours: 'Hours'`<br>`minutes: 'Minutes'`<br>`ampm: 'AM/PM'`<br>`save: 'Save'`<br>`clear: 'Clear'`<br>`close: 'Close'`<br>`}`| picker texts |

### Methods
NJ-TimePicker ships with methods that can be used to change the picker behaviour. These methods have strict input formatting which helps achive desired result

#### `picker.show()`
This is methods triggers the show event and presents the picker on-screen.

#### `picker.hide()`
This is methods triggers the hide event and hides the picker.

#### `picker.setValue(param)`
The setValue method sets the value of the picker. It take one argument which can be a string or a object
```js
// as string
picker.setValue('12'); // h-12
picker.setValue('12:45'); // h-12, m-45 
picker.setValue('12:45 am'); // h-12, m-45, meridiem-am

// as object
picker.setValue({
    hours: 12, // h-12
    minutes: 45, // m-45 
    ampm: 'am' // meridiem-am
});
```

#### `picker.getValue(param)`
The getValue method gets the value of the picker. It takes an optional argument.
```js
picker.getValue();
// result object
{
  ampm: "am"
  fullResult: "12:45 am"
  hours: "12"
  minutes: "45"
}
// with parameter
picker.getValue('hours'); // outputs hours, 12
picker.getValue('minutes'); // outputs minutes, 45
picker.getValue('am'); // outputs meridiem, am
picker.getValue('fullResult'); // outputs full string, 12:45 am
```
### Events
| Name        | Description  |
| :---------- |:-------------|
| ready       | when plugin initialization is complete |
| show        | picker is presented to the user |
| hide        | when picker is hidden |
| save        | on save button click |
| cancel      | on cancel button click |
| clear       | on clear button click |

### Theming
You can overridde the default theme by changing the following [css variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties)
```css
:root {
    /* save button */ 
    --save-button-bg: #008000;
    --save-button-fc: #ffffff;

    /* clear button */
    --clear-button-bg: #ffa500;
    --clear-button-fc: #ffffff;

    /* close button */
    --close-button-bg: #ff0000;
    --close-button-fc: #ffffff;

    /* header */ 
    --header-text-color: #ffffff;  
    --header-text-align: left;
    --header-bg: #414141;

    /* picker container */
    --picker-bg: #fff;
    --picker-selection-bg: #0000ff;
    --picker-selection-fc: #fff; 
    --picker-max-width: 400px;
    --picker-min-width: 350px;
}
```
## How to Contribute
1. Clone repo and create a new branch: `$ git checkout https://github.com/nj-coder/nj-timepicker -b name_for_new_branch`.
2. Make changes and test
3. Submit Pull Request with comprehensive description of changes

## Releases and Changes
Check out the [Releases](https://github.com/nj-coder/nj-timepicker/releases) and [Change Logs](https://github.com/nj-coder/nj-timepicker/blob/master/Changelog.md) for more information. 

## Copyright and License
[The MIT license](LICENSE.md) 


Changelog.md
# Change Log
All notable changes to this project will be documented in this file. For transparency into release cycles and in striving to maintain backward compatibility, the project is maintained under [the Semantic Versioning guidelines](https://semver.org/). 

The below format is based on [Keep a Changelog](http://keepachangelog.com/).

## [Unreleased]
### Added
- enhancement - added `autoSave` config to close the picker after selection ([4a8e091
](https://github.com/nj-coder/nj-timepicker/commit/4a8e091
)) closes [#7](https://github.com/nj-coder/nj-timepicker/issues/7)
### Changed
- updated event emitter to [NJ-Events](https://github.com/nj-coder/nj-events)
- updated build process to [gulp-better-rollup](https://www.npmjs.com/package/gulp-better-rollup)
- `setValue()` method to accept object, instead of indivual key-value pair
### Fixed
 - fix - After changing button text, ie: close: 'Cerrar', buttons become non-functional. ([571724b](https://github.com/nj-coder/nj-timepicker/commit/571724b)) closes [#8](https://github.com/nj-coder/nj-timepicker/issues/8)

## [1.3.100] - 2019-11-16
### Added
 - enhancement - auto fill value if the target element is an input ([0d00cdd](https://github.com/nj-coder/nj-timepicker/commit/0d00cdd)) closes [#4](https://github.com/nj-coder/nj-timepicker/issues/4)
 - enhancement - picker texts can be configured by providing in the `texts` configuration ([4409f99](https://github.com/nj-coder/nj-timepicker/commit/4409f99)) closes [#5](https://github.com/nj-coder/nj-timepicker/issues/5)
 - enhancement - added more variables for theming ([4409f99](https://github.com/nj-coder/nj-timepicker/commit/4409f99))
 - zoom in animation when picker is shown ([91641da](https://github.com/nj-coder/nj-timepicker/commit/91641da))

### Deprecated
- `headerText` instead use `texts` to configure the header 

## [1.2.102] - 2019-11-07 
### Added
 - `disabledHours` and `disabledMinutes` config
 - `headerText` config
### Fixed
 - An issue which returns || when save is pressed without selecting any values
 - fix - DIV renders at top of page, irregardless of page scrolling ([81de097](https://github.com/nj-coder/nj-timepicker/commit/81de097)) closes [#3](https://github.com/nj-coder/nj-timepicker/issues/3)

## [1.1.100] - 2019-09-28
### Added
 - Base component
 - Minutes configuration
 - set and get value methods
 
## [0.0.100] - 2019-08-15
### Added
 - Inital layout and configuration


src/index.js
import { Background } from './background/background';
import NJEvents from 'nj-events';
import { PickerHour as hours } from './hour/hour';
import { PickerMinute as minutes } from './minutes/minute';
import { PickerAMPM as ampm } from './ampm/ampm';
import { Header } from './header/header';
import { ActionButton as buttons } from './button/buttons';

export default class NJTimePicker {

    constructor(options = {}) {
        // checks for a valid target element
        this.checkTarget(options.targetEl || options.targetID);

        // merge options with default config
        this.config = this.extend({
            options
        });

        this.emitter = new NJEvents();
        this.pickerVisible = false;
        this.allowAutoSave = false;
        this.build(); // build the picker
    }

    // plugin default config
    get defaultConfig() {
        return {
            clickOutsideToClose: true,
            autoSave: false, // auto save if for the first time
            disabledMinutes: [],
            disabledHours: [],
            format: '12',
            texts: {
                header: '',
                hours: 'Hours',
                minutes: 'Minutes',
                close: 'Close',
                save: 'Save',
                clear: 'Clear',
                ampm: 'AM/PM'
            },
            headerText: '',
            id: Math.random().toString(36).substring(7),
            minutes: [0, 15, 30, 45],
            targetEl: null,
            targetID: null
        };
    }

    // merge user config with defaults
    extend({
        target = {},
        options,
        defaults = this.defaultConfig
    }) {
        Object.keys(defaults).forEach(key => {
            if (typeof (defaults[key]) == 'object' && key == 'texts') {
                target[key] = {};
                Object.keys(defaults[key]).forEach(e => {
                    target[key][e] = (options[key] && options[key][e]) ?
                        options[key][e] : defaults[key][e];
                });
            } else {
                target[key] = options[key] || defaults[key];
            }
        });

        return target;
    }
    // check if the target element exist in the dom
    checkTarget(el) {
        if (el instanceof HTMLElement) { // check for valid dom element
            this.targetElement = el;
        } else if (document.querySelector(`#${el}`)) { // check for valid dom element with ID targetID
            this.targetElement = document.querySelector(`#${el}`);
        } else {
            throw ('NJPicker requires a valid target element(targetEl) or target element ID(targetID)');
        }
    }

    // set the options for building the plugin
    build() {
        // attach the click to show picker
        this.targetElement.addEventListener('click', this.show.bind(this));

        // create picker wrapper div
        this.wrapper = document.createElement('div');
        this.wrapper.classList.add('nj-picker');
        this.wrapper.setAttribute('id', this.config.id);

        // init the overlay
        this.overlay = new Background();
        if (this.config.clickOutsideToClose) { // check if clickOutsideToClose is true
            this.overlay.addEventListener('click', this.hide.bind(this));
        }
        // append the overlay to the picker wrapper
        this.wrapper.append(this.overlay);

        // create the container
        this.container = document.createElement('div');
        this.container.classList.add('nj-picker-container');

        // create header
        if (this.config.headerText || this.config.texts.header) {
            this.header = new Header(this.config);
            this.container.append(this.header.element);
        }

        this.buildItems();
        this.buildButtons(); // configure picker buttons

        // attach the picker container to the wrapper
        this.wrapper.append(this.container);
        // attach the picker wrapper to the dom
        document.body.append(this.wrapper);
        setTimeout(() => {
            this.emitter.emit('ready', {}); // emit the plugin ready event
        });

        this.emitter.on('selection', () => {
            if (this.pickerVisible && this.config.autoSave && this.allowAutoSave) {
                let val = this.getValue();
                if (val.hours && val.minutes && val.ampm)
                    this.onSave();
            }
        });
    }

    // create buttons contianer
    buildButtons() {
        this.buttons = new buttons(this.config, this.emitter);
        this.emitter.on('btn-save', () => {
            // set the value of the target if its an input 
            if (this.targetElement.type == 'text' && this.targetElement.nodeName === 'INPUT') {
                this.targetElement.value = this.getValue().fullResult;
            }
            this.onSave();
        });
        this.emitter.on('btn-clear', () => {
            this.hours.resetValue(); // resets the hours
            this.minutes.resetValue(); // reset minutes
            if (this.config.format == '12') {
                this.ampm.resetValue(); // resets am-pm
            }
            this.emitter.emit('clear');
        });
        this.emitter.on('btn-close', () => {
            this.emitter.emit('close');
            this.hide();
        });
        this.container.append(this.buttons.build(this.config));
    }

    // build individual items
    buildItems() {
        // create hours contianer
        this.hours = new hours(this.config, this.emitter);
        this.container.append(this.hours.element);
        // create minutes contianer
        this.minutes = new minutes(this.config, this.emitter);
        this.container.append(this.minutes.element);
        // create ampm contianer
        if (this.config.format == '12') {
            this.ampm = new ampm(this.config, this.emitter);
            this.container.append(this.ampm.element);
        }
    }

    onSave() {
        this.emitter.emit('save', this.getValue());
        this.hide();
    }
    // get the picker value
    getValue(key) {
        let result = {
            hours: this.hours.getValue(),
            minutes: this.minutes.getValue(),
            fullResult: undefined
        };
        if (this.config.format == '12') {
            result.ampm = this.ampm.getValue();
            if (result.hours && result.minutes && result.ampm)
                result.fullResult = `${("0" + result.hours).slice(-2)}:${("0" + result.minutes).slice(-2)} ${result.ampm}`;
        } else {
            if (result.hours && result.minutes)
                result.fullResult = `${("0" + result.hours).slice(-2)}:${("0" + result.minutes).slice(-2)}`;
        }

        if (key) {
            return result[key];
        }

        return result || {};
    }

    // set picker value
    setValue(input) {
        if (typeof (input) == 'string') { // hh:mm am
            try {
                let split_1 = input.split(':');
                if (typeof (split_1[0]) != 'undefined' && split_1[0] != '') {
                    this.hours.setValue(split_1[0]);
                }
                if (typeof (split_1[1]) != 'undefined' && split_1[1] != '') {
                    let split_2 = split_1[1].split(' ');
                    this.minutes.setValue(split_2[0]);
                    this.ampm.setValue(split_2[1]);
                }
            } catch (e) {
                //
            }
        } else if (typeof (input) == 'object') {
            for (let item in input) {
                if (input[item] && this[item] && this[item].setValue) {
                    this[item].setValue(input[item].toString().toLowerCase());
                }
            }
        }
    }

    // shows the picker
    show() {
        let val = this.getValue();
        this.allowAutoSave = !(val.hours && val.minutes && val.ampm);
        this.pickerVisible = true;
        this.wrapper.classList.add('nj-picker-show');
        this.emitter.emit('show'); // emit plugin show event
    }

    // hides the picker
    hide() {
        this.pickerVisible = false;
        this.wrapper.classList.remove('nj-picker-show');
        this.emitter.emit('hide'); // emit the plugin hide event
    }

    // create an on method to mask emitter on
    on(name, value) {
        this.emitter.on(name, value);
    }

    static get version() { 
        return '1.3.118';
    }

}